File: FundComponents/SocioEconomicComponent.cs
// FUND - Climate Framework for Uncertainty, Negotiation and Distribution
// Copyright (C) 2012 David Anthoff and Richard S.J. Tol
// http://www.fund-model.org
// Licensed under the MIT license
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Esmf;
using Fund.CommonDimensions;

namespace Fund.Components.SocioEconomic
{
    public interface ISocioEconomicState
    {

        IVariable2Dimensional<Timestep, Region, double> income { get; }
        IVariable2Dimensional<Timestep, Region, double> consumption { get; }
        IVariable2Dimensional<Timestep, Region, double> ypc { get; }
        IVariable2Dimensional<Timestep, Region, double> ygrowth { get; }

        IVariable2Dimensional<Timestep, Region, double> plus { get; }
        IVariable2Dimensional<Timestep, Region, double> urbpop { get; }
        IVariable2Dimensional<Timestep, Region, double> popdens { get; }

        IVariable1Dimensional<Timestep, double> globalconsumption { get; }
        IVariable1Dimensional<Timestep, double> globalypc { get; }
        IVariable1Dimensional<Timestep, double> globalincome { get; }

        IVariable1Dimensional<Region, double> ypc90 { get; }


        IParameter2Dimensional<Timestep, Region, double> pgrowth { get; }
        IParameter2Dimensional<Timestep, Region, double> ypcgrowth { get; }
        IParameter2Dimensional<Timestep, Region, double> eloss { get; }
        IParameter2Dimensional<Timestep, Region, double> sloss { get; }
        IParameter2Dimensional<Timestep, Region, double> mitigationcost { get; }
        IParameter2Dimensional<Timestep, Region, double> area { get; }
        IParameter1Dimensional<Timestep, double> globalpopulation { get; }
        IParameter2Dimensional<Timestep, Region, double> population { get; }
        IParameter2Dimensional<Timestep, Region, double> populationin1 { get; }

        IParameter1Dimensional<Region, double> plus90 { get; }
        IParameter1Dimensional<Region, double> gdp90 { get; }
        IParameter1Dimensional<Region, double> pop90 { get; }
        IParameter1Dimensional<Region, double> urbcorr { get; }
        IParameter1Dimensional<Region, double> gdp0 { get; }


        [DefaultParameterValue(false)]
        bool runwithoutdamage { get; }

        double consleak { get; }
        double plusel { get; }

    }

    public class SocioEconomicComponent
    {
        public void Run(Clock clock, ISocioEconomicState state, IDimensions dimensions)
        {
            var s = state;
            var t = clock.Current;

            double savingsrate = 0.2;

            if (clock.IsFirstTimestep)
            {
                foreach (var r in dimensions.GetValues<Region>())
                {
                    s.income[t, r] = s.gdp0[r];
                    s.ypc[t, r] = s.income[t, r] / s.population[t, r] * 1000.0;
                    s.consumption[t, r] = s.income[t, r] * 1000000000.0 * (1.0 - savingsrate);
                }

                s.globalconsumption[t] = dimensions.GetValues<Region>().Select(r => s.consumption[t, r]).Sum();

                foreach (var r in dimensions.GetValues<Region>())
                {
                    s.ypc90[r] = s.gdp90[r] / s.pop90[r] * 1000;
                }
            }
            else
            {

                // Calculate income growth rate  
                foreach (var r in dimensions.GetValues<Region>())
                {
                    s.ygrowth[t, r] = (1 + 0.01 * s.pgrowth[t - 1, r]) * (1 + 0.01 * s.ypcgrowth[t - 1, r]) - 1;
                }

                // Calculate income  
                foreach (var r in dimensions.GetValues<Region>())
                {
                    double oldincome = s.income[t - 1, r] - ((t >= Timestep.FromSimulationYear(40)) && !s.runwithoutdamage ? s.consleak * s.eloss[t - 1, r] / 10.0 : 0);

                    s.income[t, r] = (1 + s.ygrowth[t, r]) * oldincome - s.mitigationcost[t - 1, r];
                }

                // Check for unrealistic values
                foreach (var r in dimensions.GetValues<Region>())
                {
                    if (s.income[t, r] < 0.01 * s.population[t, r])
                        s.income[t, r] = 0.1 * s.population[t, r];

                }

                foreach (var r in dimensions.GetValues<Region>())
                {
                    s.ypc[t, r] = s.income[t, r] / s.population[t, r] * 1000.0;
                }

                var totalConsumption = 0.0;
                foreach (var r in dimensions.GetValues<Region>())
                {
                    s.consumption[t, r] = Math.Max(
                        s.income[t, r] * 1000000000.0 * (1.0 - savingsrate) - (s.runwithoutdamage ? 0.0 : (s.eloss[t - 1, r] + s.sloss[t - 1, r]) * 1000000000.0),
                      0.0);

                    totalConsumption = totalConsumption + s.consumption[t, r];
                }
                s.globalconsumption[t] = totalConsumption;

                foreach (var r in dimensions.GetValues<Region>())
                {
                    s.plus[t, r] = s.plus90[r] * Math.Pow(s.ypc[t, r] / s.ypc90[r], s.plusel);

                    if (s.plus[t, r] > 1)
                        s.plus[t, r] = 1.0;
                }

                foreach (var r in dimensions.GetValues<Region>())
                {
                    s.popdens[t, r] = s.population[t, r] / s.area[t, r] * 1000000.0;
                }

                foreach (var r in dimensions.GetValues<Region>())
                {
                    s.urbpop[t, r] = (0.031 * Math.Sqrt(s.ypc[t, r]) - 0.011 * Math.Sqrt(s.popdens[t, r])) / (1.0 + 0.031 * Math.Sqrt(s.ypc[t, r]) - 0.011 * Math.Sqrt(s.popdens[t, r]))
                        / (1 + s.urbcorr[r] / (1 + 0.001 * Math.Pow(Convert.ToDouble(t.Value) - 40.0, 2))); // (* DA: urbcorr needs to be changed to a function if this is to be made uncertain *)

                }

                s.globalincome[t] = dimensions.GetValues<Region>().Select(r => s.income[t, r]).Sum();

                s.globalypc[t] = dimensions.GetValues<Region>().Select(r => s.income[t, r] * 1000000000.0).Sum() /
                    dimensions.GetValues<Region>().Select(r => s.populationin1[t, r]).Sum();
            }
        }
    }
}

File: src/components/SocioEconomicComponent.jl
@defcomp socioeconomic begin
	regions = Index()

	income = Variable(index=[time,regions])
	consumption = Variable(index=[time,regions])
	ypc = Variable(index=[time,regions])
	ygrowth = Variable(index=[time,regions])

	plus = Variable(index=[time,regions])
	urbpop = Variable(index=[time,regions])
	popdens = Variable(index=[time,regions])

	globalconsumption = Variable(index=[time])
	globalypc = Variable(index=[time])
	globalincome = Variable(index=[time])

	ypc90 = Variable(index=[regions])


	pgrowth             = Parameter(index=[time,regions])
	ypcgrowth           = Parameter(index=[time,regions])
	eloss               = Parameter(index=[time,regions])
	sloss               = Parameter(index=[time,regions])
	mitigationcost      = Parameter(index=[time,regions])
	area                = Parameter(index=[time,regions])
	globalpopulation    = Parameter(index=[time])
	population          = Parameter(index=[time,regions])
	populationin1       = Parameter(index=[time,regions])

	plus90  = Parameter(index=[regions])
	gdp90   = Parameter(index=[regions])
	pop90   = Parameter(index=[regions])
	urbcorr = Parameter(index=[regions])
	gdp0    = Parameter(index=[regions])


	runwithoutdamage = Parameter{Bool}(default=false)
	consleak    = Parameter(default=0.25)
	plusel      = Parameter(default=0.25)
    savingsrate = Parameter(default=0.2)

    function run_timestep(p, v, d, t)

        if is_first(t)
            for r in d.regions
                v.income[t, r] = p.gdp0[r]
                v.ypc[t, r] = v.income[t, r] / p.population[t, r] * 1000.0
                v.consumption[t, r] = v.income[t, r] * 1000000000.0 * (1.0 - p.savingsrate)
            end

            v.globalconsumption[t] = sum(v.consumption[t,:])

            for r in d.regions
                v.ypc90[r] = p.gdp90[r] / p.pop90[r] * 1000
            end
        else

            # Calculate income growth rate
            for r in d.regions
                v.ygrowth[t, r] = (1 + 0.01 * p.pgrowth[t - 1, r]) * (1 + 0.01 * p.ypcgrowth[t - 1, r]) - 1
            end

            # Calculate income
            for r in d.regions
                oldincome = v.income[t - 1, r] - (gettime(t) >= 1990 && !p.runwithoutdamage ? p.consleak * p.eloss[t - 1, r] / 10.0 : 0)

                v.income[t, r] = (1 + v.ygrowth[t, r]) * oldincome - p.mitigationcost[t - 1, r]
            end

            # Check for unrealistic values
            for r in d.regions
                if v.income[t, r] < 0.01 * p.population[t, r]
                    v.income[t, r] = 0.1 * p.population[t, r]
                end
            end

            for r in d.regions
                v.ypc[t, r] = v.income[t, r] / p.population[t, r] * 1000.0
            end

            for r in d.regions
                v.consumption[t, r] = max(v.income[t, r] * 1000000000.0 * (1.0 - p.savingsrate) - (p.runwithoutdamage ? 0.0 :   (p.eloss[t - 1, r] + p.sloss[t - 1, r]) * 1000000000.0), 0.0)
            end
            v.globalconsumption[t] = sum(v.consumption[t,:])

            for r in d.regions
                v.plus[t, r] = p.plus90[r] * (v.ypc[t, r] / v.ypc90[r])^p.plusel

                if v.plus[t, r] > 1
                    v.plus[t, r] = 1.0
                end
            end

            for r in d.regions
                v.popdens[t, r] = p.population[t, r] / p.area[t, r] * 1000000.0
            end

            for r in d.regions
                # ERROR This doesn't make sense for t < 40
                v.urbpop[t, r] = (0.031 * sqrt(v.ypc[t, r]) - 0.011 * sqrt(v.popdens[t, r])) / (1.0 + 0.031 * sqrt(v.ypc[t,r]) - 0.011 * sqrt(v.popdens[t, r])) / (1 + p.urbcorr[r] / (1 + 0.001 * (gettime(t) - 1990)^2.))
                # DA: urbcorr needs to be changed to a function if this is to be made uncertain
            end
            
            v.globalincome[t] = sum(v.income[t,:])

            v.globalypc[t] = sum(v.income[t,:] .* 1000000000.0) / sum(p.populationin1[t,:])
        end
    end
end